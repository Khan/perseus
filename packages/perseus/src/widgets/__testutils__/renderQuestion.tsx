import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
