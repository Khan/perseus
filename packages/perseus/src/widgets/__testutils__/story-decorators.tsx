import * as React from "react";

export const mobileDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile">
        <Story />
    </div>
);

export const articleDecorator = (Story) => (
    <div className="framework-perseus perseus-article">
        <Story />
    </div>
);

export const mobileArticleDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile perseus-article">
        <Story />
    </div>
);

export const rtlDecorator = (Story) => (
    <div style={{direction: "rtl"}}>
        <Story />
    </div>
);
