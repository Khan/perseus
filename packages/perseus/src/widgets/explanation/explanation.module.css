@layer reset, shared, legacy;

@layer legacy {

    /*
      Special styling is needed to fit the button in a block of text without throwing off the line spacing.
      While the button is not normally included in a block of text, it needs to be able to accommodate such a case.
    */
    .button-style-overrides {
        height: auto;
        line-height: inherit;
        margin-left: -2px; /* transfer space on the left side to the padding */
        margin-right: 2px; /* adds space between the right border and any text that follows */
        padding-left: 2px; /* adds space between the left border and the button text */
    }

    .button-style-overrides > span {
        font-size: 18px;
        line-height: inherit;
        text-align: left;
        /*
          The following property adjusts the large space between the button text and the caret icon.
          Since we are unable to adjust the styling of the icon (where the extra space exists),
          we are adjusting it on the text label by using a negative margin.
        */
        margin-right: -6px;
        white-space: normal;
    }

    .content {
        border-left: 0px solid #ccc;
        display: inline-grid;
        position: relative;
    }

    .content-collapsed {
        grid-template-columns: 0fr;
        grid-template-rows: 0fr;
        margin-bottom: 0;
        margin-top: 0;
        min-width: 0;
        padding-bottom: 0;
        visibility: hidden;
    }

    .content-expanded {
        border-left-width: 5px;
        grid-template-columns: 1fr;
        grid-template-rows: 1fr;
        margin-left: -23px;
        min-width: 100%;
        padding-left: 23px;
        padding-bottom: 10px;
        visibility: visible;
        /*
          Note: we still use arrow height as the vertical margin, even on
          desktop when there is no arrow, but it's good enough.
        */
        margin-bottom: 14px;
        margin-top: 14px;
    }

    .content-wrapper {
        overflow: hidden;
    }

    .transition-collapsed {
        transition: all 0.25s step-end, grid-template-rows 0.25s, margin-top 0.25s, margin-bottom 0.25s, padding-bottom 0.25s;
    }

    .transition-expanded {
        transition: grid-template-rows 0.5s, margin-top 0.5s, margin-bottom 0.5s, padding-bottom 0.5s;
    }

}
