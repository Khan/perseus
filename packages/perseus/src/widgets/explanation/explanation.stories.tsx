import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";
import {generateTestPerseusItem} from "../../util/test-utils";

import {
    ipsumExample,
    question1,
    question2,
    wideButton,
} from "./explanation.testdata";

import type {Meta, StoryObj} from "@storybook/react";

const meta: Meta = {
    title: "Perseus/Widgets/Explanation",
    component: ServerItemRendererWithDebugUI,
};
export default meta;

type Story = StoryObj<typeof ServerItemRendererWithDebugUI>;

export const Question1: Story = {
    args: {item: generateTestPerseusItem({question: question1})},
};

export const Question2: Story = {
    args: {item: generateTestPerseusItem({question: question2})},
};

export const IpsumExample: Story = {
    args: {item: generateTestPerseusItem({question: ipsumExample})},
};

export const WideButton: Story = {
    args: {item: generateTestPerseusItem({question: wideButton})},
};
