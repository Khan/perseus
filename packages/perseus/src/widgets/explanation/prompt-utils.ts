import {WidgetType} from "../../prompt-types";

import type explanation from "./explanation";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

export type ExplanationPromptJSON = {
    type: WidgetType;
    showPrompt: PropsFor<typeof explanation.widget>["showPrompt"];
    hidePrompt: PropsFor<typeof explanation.widget>["hidePrompt"];
    explanation: PropsFor<typeof explanation.widget>["explanation"];
};

export const getPromptJSON = (
    renderProps: PropsFor<typeof explanation.widget>,
): ExplanationPromptJSON => {
    return {
        type: WidgetType.EXPLANATION,
        showPrompt: renderProps.showPrompt || "",
        hidePrompt: renderProps.hidePrompt || "",
        explanation: renderProps.explanation || "",
    };
};
