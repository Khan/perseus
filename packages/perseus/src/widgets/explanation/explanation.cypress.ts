import renderQuestion from "../../../../../testing/render-question-with-cypress";
import {cypressTestDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import * as Perseus from "../../index";

import {ipsumExample} from "./explanation.testdata";

//  NOTE: The regression tests in this file use Cypress because they are intended to validate styling that is applied.
//        Since React Testing Library isn't applying the CSS to the elements,
//          we can't use Jest to verify that some keyboard interactions work properly.

describe("Explanation Widget", () => {
    const postContentLinkText = "the system J-25";
    const contentLinkOne = "synchronic distortion";
    const contentLinkTwo = "selective molecular polarization";

    beforeEach(() => {
        Dependencies.setDependencies(cypressTestDependencies);
        Perseus.init();
    });

    it("prevents interacting with actionable items within content when COLLAPSED (initial state)", () => {
        // Arrange
        renderQuestion(ipsumExample);
        cy.get("button[aria-expanded='false'][aria-controls]").focus();

        // Act - verify tab order (forwards)
        cy.focused().should("have.text", "Explanation"); // Verify we are on the widget's button.
        cy.realPress("Tab");
        cy.focused().should("have.text", postContentLinkText);

        // Act - verify tab order (backwards)
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", "Explanation");
    });

    it("allows interacting with actionable items within content when EXPANDED", () => {
        // NOTE: This test ensures that the CSS that controls keyboard access doesn't regress.
        //       It also ensures that any JavaScript event handling doesn't interfere with expected keyboard navigation.

        // Arrange
        renderQuestion(ipsumExample);
        cy.get("button[aria-expanded='false'][aria-controls]").focus();
        cy.focused().should("have.text", "Explanation"); // Verify we are on the widget's button.
        cy.realPress("Enter"); // Expand content

        // Act - verify tab order (forwards)
        cy.realPress("Tab");
        cy.focused().should("have.text", contentLinkOne);
        cy.realPress("Tab");
        cy.focused().should("have.text", contentLinkTwo);
        cy.realPress("Tab");
        cy.focused().should("have.text", postContentLinkText);

        // Act - verify tab order (backwards)
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", contentLinkTwo);
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", contentLinkOne);
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", "Hide"); // "Hide" is the new text in the widget's button.
    });

    it("prevents interacting with actionable items within content when COLLAPSED (after toggle)", () => {
        // NOTE: This test ensures that interaction with the widget's button doesn't regress.
        //       The test is similar to the first "COLLAPSED" test,
        //          but while that one tests the initial state of the widget,
        //          this one verifies that toggling doesn't introduce/remove anything that would interfere with
        //          expected keyboard navigation.

        // Arrange
        renderQuestion(ipsumExample);
        cy.get("button[aria-expanded='false'][aria-controls]").focus();
        cy.focused().should("have.text", "Explanation"); // Verify we are on the widget's button.
        cy.realPress("Enter"); // Expand content

        // Act - verify tab order (forwards)
        cy.realPress("Tab");
        cy.focused().should("have.text", contentLinkOne);

        // Act - verify tab order (backwards)
        cy.realPress(["Shift", "Tab"]);
        cy.realPress("Enter"); // Collapse content
        cy.waitUntil(() =>
            cy
                .focused()
                .siblings()
                .first()
                .should("have.attr", "aria-hidden", "true")
                .should("have.css", "visibility", "hidden"),
        );
        cy.realPress("Tab");
        cy.focused().should("have.text", postContentLinkText);
    });
});
