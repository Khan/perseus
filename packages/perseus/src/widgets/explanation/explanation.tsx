import {linterContextDefault} from "@khanacademy/perseus-linter";
import Button from "@khanacademy/wonder-blocks-button";
import {Id, View} from "@khanacademy/wonder-blocks-core";
import caretDown from "@phosphor-icons/core/regular/caret-down.svg";
import caretUp from "@phosphor-icons/core/regular/caret-up.svg";
import * as React from "react";
import _ from "underscore";

import {PerseusI18nContext} from "../../components/i18n-context";
import * as Changeable from "../../mixins/changeable";
import Renderer from "../../renderer";
import {getPromptJSON as _getPromptJSON} from "../../widget-ai-utils/explanation/explanation-ai-utils";
import scoreNoop from "../__shared__/score-noop";

import styles from "./explanation.module.css";

import type {PerseusExplanationWidgetOptions} from "../../perseus-types";
import type {Widget, WidgetExports, WidgetProps} from "../../types";
import type {ExplanationPromptJSON} from "../../widget-ai-utils/explanation/explanation-ai-utils";
import type {PerseusExplanationWidgetOptions} from "@khanacademy/perseus-core";

type RenderProps = PerseusExplanationWidgetOptions; // transform = _.identity

type Props = WidgetProps<RenderProps>;

type DefaultProps = {
    showPrompt: Props["showPrompt"];
    hidePrompt: Props["hidePrompt"];
    explanation: Props["explanation"];
    widgets: Props["widgets"];
    linterContext: Props["linterContext"];
};

type State = {
    expanded: boolean;
};

function mediaQueryIsMatched(mediaQuery: string): boolean {
    if (typeof window.matchMedia !== "function") {
        return false;
    }
    return window.matchMedia(mediaQuery).matches;
}

class Explanation extends React.Component<Props, State> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    static defaultProps: DefaultProps = {
        showPrompt: "Explain",
        hidePrompt: "Hide explanation",
        explanation: "explanation goes here\n\nmore explanation",
        widgets: {},
        linterContext: linterContextDefault,
    };

    // this just helps with TS weak typing when a Widget
    // doesn't implement any Widget methods
    isWidget = true as const;

    state: State = {
        expanded: false,
    };
    _mounted: boolean = false;

    componentDidMount() {
        this._mounted = true;
    }

    componentWillUnmount() {
        this._mounted = false;
    }

    change: (arg1: any, arg2: any, arg3: any) => any = (...args) => {
        return Changeable.change.apply(this, args);
    };

    _onClick: () => void = () => {
        this.setState({
            expanded: !this.state.expanded,
        });
        this.props.trackInteraction();
    };

    getPromptJSON(): ExplanationPromptJSON {
        return _getPromptJSON(this.props);
    }

    render(): React.ReactNode {
        const promptText = this.state.expanded
            ? this.props.hidePrompt
            : this.props.showPrompt;

        const caretIcon = this.state.expanded ? caretUp : caretDown;

        const allowTransition =
            this._mounted &&
            mediaQueryIsMatched("(prefers-reduced-motion: no-preference)");

        const contentStyling = [
            styles.content,
            this.state.expanded
                ? styles.contentExpanded
                : styles.contentCollapsed,
            allowTransition &&
                (this.state.expanded
                    ? styles.transitionExpanded
                    : styles.transitionCollapsed),
        ];

        return (
            <Id>
                {(contentId) => (
                    <>
                        <Button
                            aria-expanded={this.state.expanded}
                            aria-controls={contentId}
                            endIcon={caretIcon}
                            kind="tertiary"
                            onClick={this._onClick}
                            size="small"
                            className={styles.buttonStyleOverrides}
                        >
                            {promptText}
                        </Button>

                        <View
                            id={contentId}
                            className={contentStyling.join(" ")}
                            aria-hidden={!this.state.expanded}
                            testId="content-container"
                        >
                            <View className={styles.contentWrapper}>
                                <Renderer
                                    apiOptions={this.props.apiOptions}
                                    content={this.props.explanation}
                                    widgets={this.props.widgets}
                                    linterContext={this.props.linterContext}
                                    strings={this.context.strings}
                                />
                            </View>
                        </View>
                    </>
                )}
            </Id>
        );
    }
}

export default {
    name: "explanation",
    displayName: "Explanation",
    accessible: true,
    defaultAlignment: "inline",
    widget: Explanation,
    transform: _.identity,
    isLintable: true,
    // TODO: things that aren't interactive shouldn't need scoring functions
    scorer: () => scoreNoop(),
} satisfies WidgetExports<typeof Explanation>;
