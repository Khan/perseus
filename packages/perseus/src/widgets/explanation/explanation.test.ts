import {act, screen} from "@testing-library/react";
import {userEvent as userEventLib} from "@testing-library/user-event";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import * as Changeable from "../../mixins/changeable";
import {renderQuestion} from "../__testutils__/renderQuestion";

import ExplanationWidgetExports from "./explanation";
import {question1} from "./explanation.testdata";

import type {UserEvent} from "@testing-library/user-event";

describe("Explanation", function () {
    let userEvent: UserEvent;

    // NOTE: Since the visibility of an element is controlled by CSS,
    //          the only way that we can verify in RTL that an element is visible or not (expanded/collapsed)
    //          is by checking the classes that are applied to the wrapper of that element.
    //       Therefore, we are checking the wrapper element by its data-testid instead of the normal
    //          getByRole or getByText functions.
    //       The same is true for the animation tests found in this file.

    const verifyExpandCollapseState = (
        buttonText: string,
        isExpanded: boolean,
    ) => {
        // Widget Button state
        const widgetButton = screen.getByRole("button", {name: buttonText});
        expect(widgetButton).toHaveAttribute(
            "aria-expanded",
            String(isExpanded),
        );

        // Content container state
        const contentContainer = screen.getByTestId("content-container");
        expect(contentContainer).toHaveAttribute(
            "aria-hidden",
            String(!isExpanded),
        );

        const expectedClass = isExpanded
            ? "contentExpanded"
            : "contentCollapsed";
        expect(contentContainer.className).toContain(expectedClass);

        const excludedClass = isExpanded
            ? "contentCollapsed"
            : "contentExpanded";
        expect(contentContainer.className).not.toContain(excludedClass);
    };

    const getMatchMediaMockFn = (doesMatch: boolean, mediaQuery?: string) => {
        return (query) => ({
            matches: (mediaQuery ?? query) === query ? doesMatch : !doesMatch,
            media: query,
            onchange: null,
            addEventListener: jest.fn(),
            addListener: jest.fn(),
            removeListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        });
    };

    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange and Act
        const {container} = renderQuestion(question1);

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should snapshot when expanded", async () => {
        // Arrange
        const {container} = renderQuestion(question1);

        // Act
        await userEvent.click(screen.getByRole("button"));

        // Assert
        // The only real difference between expanded and not expanded is the
        //     classes and aria that are applied.
        expect(container).toMatchSnapshot("expanded");
    });

    it("can be expanded and collapsed with a mouse click", async function () {
        // Arrange
        renderQuestion(question1);

        // Verify initial state
        verifyExpandCollapseState("Explanation", false);

        // Act - expand with a click
        await userEvent.click(
            screen.getByRole("button", {name: "Explanation"}),
        );

        // Assert - elements have attributes changed that represent an expanded state
        verifyExpandCollapseState("Hide explanation!", true);

        // Act - collapse with a click
        await userEvent.click(
            screen.getByRole("button", {name: "Hide explanation!"}),
        );

        // Assert - elements have attributes reset to represent a collapsed state
        verifyExpandCollapseState("Explanation", false);
    });

    it("can be expanded and collapsed with the keyboard - Enter key", async function () {
        // Arrange
        renderQuestion(question1);

        // Verify initial state
        verifyExpandCollapseState("Explanation", false);

        // Act - expand with the enter key
        act(() => screen.getByRole("button", {name: "Explanation"}).focus());
        await userEvent.keyboard("{Enter}");

        // Assert - elements have attributes changed that represent an expanded state
        verifyExpandCollapseState("Hide explanation!", true);

        // Act - collapse with the enter key
        act(() =>
            screen.getByRole("button", {name: "Hide explanation!"}).focus(),
        );
        await userEvent.keyboard("{Enter}");

        // Assert - elements have attributes reset to represent a collapsed state
        verifyExpandCollapseState("Explanation", false);
    });

    it("can be expanded and collapsed with the keyboard - Space bar", async function () {
        // Arrange
        renderQuestion(question1);

        // Verify initial state
        verifyExpandCollapseState("Explanation", false);

        // Act - expand with a space bar
        act(() => screen.getByRole("button", {name: "Explanation"}).focus());
        await userEvent.keyboard(" ");

        // Assert - elements have attributes changed that represent an expanded state
        verifyExpandCollapseState("Hide explanation!", true);

        // Act - collapse with a space bar
        act(() =>
            screen.getByRole("button", {name: "Hide explanation!"}).focus(),
        );
        await userEvent.keyboard(" ");

        // Assert - elements have attributes reset to represent a collapsed state
        verifyExpandCollapseState("Explanation", false);
    });

    it("uses transitions when it is expanded/collapsed", async () => {
        // Arrange
        jest.spyOn(window, "matchMedia").mockImplementation(
            getMatchMediaMockFn(
                true,
                "(prefers-reduced-motion: no-preference)",
            ),
        );
        renderQuestion(question1);

        // Act - expand
        await userEvent.click(
            screen.getByRole("button", {name: "Explanation"}),
        );

        // Assert - transition when revealing
        expect(screen.getByTestId("content-container").className).toContain(
            "transitionExpanded",
        );

        // Act - collapse
        await userEvent.click(
            screen.getByRole("button", {name: "Hide explanation!"}),
        );

        // Assert - transition when concealing
        expect(screen.getByTestId("content-container").className).toContain(
            "transitionCollapsed",
        );
    });

    it("does NOT use transitions when matchMedia is not available", async () => {
        // e.g. during SSR in webapp
        const fakeWindow = Object.create(window);
        fakeWindow.matchMedia = undefined;
        jest.spyOn(window, "window", "get").mockReturnValue(fakeWindow as any);
        renderQuestion(question1);

        // Act - expand
        await userEvent.click(
            screen.getByRole("button", {name: "Explanation"}),
        );

        // Assert - don't transition when revealing
        expect(screen.getByTestId("content-container").className).not.toContain(
            "transitionExpanded",
        );
    });

    it("does NOT use transitions when the user prefers reduced motion", async () => {
        // Arrange
        jest.spyOn(window, "matchMedia").mockImplementation(
            getMatchMediaMockFn(
                false,
                "(prefers-reduced-motion: no-preference)",
            ),
        );
        renderQuestion(question1);

        // Act - expand
        await userEvent.click(
            screen.getByRole("button", {name: "Explanation"}),
        );

        // Assert - don't transition when revealing
        expect(screen.getByTestId("content-container").className).not.toContain(
            "transitionExpanded",
        );

        // Act - collapse
        await userEvent.click(
            screen.getByRole("button", {name: "Hide explanation!"}),
        );

        // Assert - don't transition when concealing
        expect(screen.getByTestId("content-container").className).not.toContain(
            "transitionCollapsed",
        );
    });

    it("communicates changes to its parent by using the provided 'onChange' callback", () => {
        // Arrange
        // @ts-expect-error // Argument of type {onChange: () => void;} is not assignable to parameter of type Props | Readonly<Props>
        const widget = new ExplanationWidgetExports.widget({
            onChange: () => {},
        });
        const callbackMock = jest.fn();
        const changeMock = jest
            .spyOn(Changeable, "change")
            .mockImplementation(() => {});

        // Act - call the widget's "change" function
        widget.change("foo", "bar", callbackMock);

        // Assert
        expect(changeMock.mock.contexts[0]).toEqual(widget);
        expect(changeMock).toHaveBeenCalledWith("foo", "bar", callbackMock);
    });

    it("should return an empty object for getUserInput()", async () => {
        // Arrange
        const {renderer} = renderQuestion(question1);

        // Act
        const userInput = renderer.getUserInput();

        // Assert
        expect(userInput).toMatchInlineSnapshot(`
            [
              {},
            ]
        `);
    });

    it("should return a zero score for simpleValidate()", async () => {
        // Arrange
        const {renderer} = renderQuestion(question1);

        // Act
        const score = renderer.scoreWidgets();

        // Assert
        expect(score).toMatchInlineSnapshot(`
            {
              "explanation 1": {
                "earned": 0,
                "message": null,
                "total": 0,
                "type": "points",
              },
            }
        `);
    });

    describe("static validate", () => {
        it("should always return 0 points", async () => {
            const result = ExplanationWidgetExports.widget.validate(
                {},
                question1.widgets["explanation 1"].options,
            );

            expect(result).toMatchInlineSnapshot(`
                {
                  "earned": 0,
                  "message": null,
                  "total": 0,
                  "type": "points",
                }
            `);
        });
    });
});
