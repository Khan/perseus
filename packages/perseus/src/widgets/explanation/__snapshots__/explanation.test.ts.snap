// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explanation should snapshot when expanded: expanded 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <button
            aria-controls=":r1:"
            aria-disabled="false"
            aria-expanded="true"
            class="button_vr44p2-o_O-shared_lwskrm-o_O-default_1hl5pu8-o_O-small_14crccx-o_O-inlineStyles_1s8anjv"
            role="button"
            type="button"
          >
            <span
              class="text_f1191h-o_O-LabelSmall_hl4zs1-o_O-text_6kf3xs-o_O-textWithFocus_e296pg-o_O-hover_1nzvkun-o_O-inlineStyles_egvfd8"
            >
              Hide explanation!
            </span>
            <div
              class="default_xu2jcg-o_O-iconWrapper_1nxzm3x-o_O-endIconWrapper_l32m71-o_O-endIconWrapperTertiary_10fyztj-o_O-iconWrapperSecondaryHovered_klcp8i"
            >
              <span
                aria-hidden="true"
                class="svg_1q6jc65-o_O-icon_1c9fkuf-o_O-inlineStyles_1troscl"
              />
            </div>
          </button>
          <div
            aria-hidden="false"
            class="default_xu2jcg-o_O-content_1pprw9s-o_O-contentExpanded_1ojyq0m"
            data-testid="content-container"
            id=":r1:"
          >
            <div
              class="default_xu2jcg-o_O-contentWrapper_y9ev9r"
            >
              <div
                class="perseus-renderer perseus-renderer-responsive"
              >
                <div
                  class="paragraph"
                  data-perseus-paragraph-index="0"
                >
                  <div
                    class="paragraph"
                  >
                    This is an explanation
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot: initial render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <button
            aria-controls=":r0:"
            aria-disabled="false"
            aria-expanded="false"
            class="button_vr44p2-o_O-shared_lwskrm-o_O-default_1hl5pu8-o_O-small_14crccx-o_O-inlineStyles_1s8anjv"
            role="button"
            type="button"
          >
            <span
              class="text_f1191h-o_O-LabelSmall_hl4zs1-o_O-text_6kf3xs-o_O-textWithFocus_e296pg-o_O-inlineStyles_egvfd8"
            >
              Explanation
            </span>
            <div
              class="default_xu2jcg-o_O-iconWrapper_1nxzm3x-o_O-endIconWrapper_l32m71-o_O-endIconWrapperTertiary_10fyztj"
            >
              <span
                aria-hidden="true"
                class="svg_1q6jc65-o_O-icon_1c9fkuf-o_O-inlineStyles_urgnaq"
              />
            </div>
          </button>
          <div
            aria-hidden="true"
            class="default_xu2jcg-o_O-content_1pprw9s-o_O-contentCollapsed_j2403j"
            data-testid="content-container"
            id=":r0:"
          >
            <div
              class="default_xu2jcg-o_O-contentWrapper_y9ev9r"
            >
              <div
                class="perseus-renderer perseus-renderer-responsive"
              >
                <div
                  class="paragraph"
                  data-perseus-paragraph-index="0"
                >
                  <div
                    class="paragraph"
                  >
                    This is an explanation
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;
